Docs: Subgraph Isomorphism (#2000)
oneapi-src/oneDAL: docs/source/onedal/algorithms/graph/subgraph-isomorphism.rst

.. default-domain:: cpp

.. _alg_subgraph_isomorphism:

====================
Subgraph isomorphism
====================

.. include::  ../../../includes/graph/subgraph-isomorphism-introduction.rst

------------------------
Mathematical formulation
------------------------

.. _subgraph_isomorphism_compute:

Subgraphs definition
--------------------

A graph :math:`H = (V'; E')` is called a subgraph of graph :math:`G = (V; E)` if 
:math:`V' \subseteq V; E' \subseteq E` and :math:`V'` contains all the endpoints of all the 
edges in :math:`E'` [Gross2014]_.

Further we denote the :capterm:`induced subgraph on the vertex set<Induced subgraph on the vertex set>` 
as **induced** subgraph, the :capterm:`induced subgraph on the edge set<Induced subgraph on the edge set>`
as **non-induced** subgraph.

Computing
---------

Given two graphs :math:`G` and :math:`H`, the problem is to determine whether graph :math:`G` contains
a subgraph isomorphic to graph :math:`H` and find the exact mapping of subgraph :math:`H` in
graph :math:`G`.

:math:`G` is called **target** graph, :math:`H` is called **pattern** graph.

> Mapping is a bijection or one-to-one correspondence between vertices of :math:`H` and subgraph of 

Reviewer: ```suggestion
Mapping is a bijection or one-to-one correspondence between vertices of :math:`H` and a subgraph of 
```

> graph :math:`G`. Every existing edge corresponds two vertices adjacency relationship, non-existing edge 
> corresponds two vertices non-adjacency relationship. Induced kind of subgraph isomorphims preserves
> adjacency and non-adjacency relationship between vertices, when non-induced kind preserveds only 
> adjacency relationship.

Reviewer: maybe like this?

```suggestion
graph :math:`G`. Two vertices are adjacent if there is an existing edge between them, and non-adjacent otherwise.
Induced subgraph isomorphism preserves both adjacency and non-adjacency relationships between vertices,
while non-induced subgraph isomorphism preserves only adjacency relationship.
```

note that you have a typo in isomorphism

Reviewer: if you want to have something closer to your original text, then maybe this is also ok:
```suggestion
graph :math:`G`. Each existing edge denotes adjacency relationship between two vertices, while each non-existing edge 
denotes non-adjacency relationship between two vertices. Induced kind of subgraph isomorphism preserves
adjacency and non-adjacency relationship between vertices, while non-induced kind preserves only 
adjacency relationship.
```

.. *
.. * Unless required by applicable law or agreed to in writing, software
.. * distributed under the License is distributed on an "AS IS" BASIS,
.. * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
.. * See the License for the specific language governing permissions and
.. * limitations under the License.
.. *******************************************************************************/

.. default-domain:: cpp

.. _alg_subgraph_isomorphism:

====================
Subgraph isomorphism
====================

.. include::  ../../../includes/graph/subgraph-isomorphism-introduction.rst

------------------------
Mathematical formulation
------------------------

.. _subgraph_isomorphism_compute:

Subgraphs definition
--------------------

A graph :math:`H = (V'; E')` is called a subgraph of graph :math:`G = (V; E)` if 
:math:`V' \subseteq V; E' \subseteq E` and :math:`V'` contains all the endpoints of all the 
edges in :math:`E'` [Gross2014]_.

Further we denote the :capterm:`induced subgraph on the vertex set<Induced subgraph on the vertex set>` 
as **induced** subgraph, the :capterm:`induced subgraph on the edge set<Induced subgraph on the edge set>`

> as **non-induced** subgraph.

Author: We use terms from book Gross2014, but they are too long to use it in API, so we utilize common-used synonyms in API and I think it is important to emphasize its usage by bold font.

Computing
---------

Given two graphs :math:`G` and :math:`H`, the problem is to define whether graph :math:`G` contains
a subgraph which is isomorphic to graph :math:`H` and find exact mapping of subgraph :math:`H` in
graph :math:`G`.

> Mapping is bijection or one-to-one correspondence between vertices of :math:`H` and subgraph of 

Reviewer: ```suggestion
Mapping is a bijection or one-to-one correspondence between vertices of :math:`H` and a subgraph of 
```

graph :math:`G` with correspond vertex adjacensy relationship preserving for non-induced subgraph 

> and adjacensy and non-adjacensy preserving relationship for induced case. :math:`G` is called

Reviewer: you have a typo: `adjacensy` -> `adjacency`

also this part of the sentence is not clear. I think it's better to make it a separate sentence instead and rephrase

Author: will be rephrased

> **target** graph, :math:`H` is called **pattern** graph.

Reviewer: It's better not to use `**` markup because it doesn't have any semantic meaning

for the definition of these terms it's better to use rst markup (glossary) so that you could reference it later

Author: It is local subgraph isomorphism terms. I think it is important to emphasise here by bold font.

> .. figure:: images/subgraph-isomorphism-target-pattern.png

Reviewer: add a `.. rubric:: Example` before this figure

    :width: 600
    :align: center
    :alt:

> The only mapping for subgraph :math:`H` in graph :math:`G` from picture above in induced case is: 
> [3, 0, 1, 4, 2, 5]. Which means that:
> 
> * pattern vertex with id 0 is mapped on vertex in target graph with id 3; 
> * pattern vertex with id 1 is mapped on vertex in target graph with id 0;
> * pattern vertex with id 2 is mapped on vertex in target graph with id 1;
> * pattern vertex with id 3 is mapped on vertex in target graph with id 4.
> * pattern vertex with id 4 is mapped on vertex in target graph with id 2.
> * pattern vertex with id 5 is mapped on vertex in target graph with id 5.
> 
> 
> Mappings for subgraph :math:`H` in graph :math:`G` in non-induced case are: [3, 0, 1, 4, 2, 5] and [3, 6, 1, 4, 2, 5].

Reviewer: ```suggestion
For the example above, the mappings for subgraph :math:`H` in graph :math:`G` are:

- Induced: [3, 0, 1, 4, 2, 5]
- Non-induced: [3, 0, 1, 4, 2, 5] and [3, 6, 1, 4, 2, 5]

The notation [3, 0, 1, 4, 2, 5] means that:

* pattern vertex with id 0 is mapped on vertex in target graph with id 3
* pattern vertex with id 1 is mapped on vertex in target graph with id 0
* pattern vertex with id 2 is mapped on vertex in target graph with id 1
* pattern vertex with id 3 is mapped on vertex in target graph with id 4
* pattern vertex with id 4 is mapped on vertex in target graph with id 2
* pattern vertex with id 5 is mapped on vertex in target graph with id 5
```

I'd suggest to rewrite the list that explains the mappings. For example, like this:

```
a vertex from the pattern graph with id=0 is mapped on a vertex in the target graph with id=3
```

or `onto a vertex`?

Author: both are ok
